File: SrcCPP/sequence.cpp
// Copyright 2004 Laurent GUEGUEN

// This file is part of Sarment.

// Sarment is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.

// Sarment is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Sarment; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#include "sequence.h"
#include "partition.h"
#include "lexique.h"

#include <iostream>
#include <ctype.h>
#include <stdlib.h>

using namespace std;

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
////////////////////////////////////////////////////////////////

Csequence::Csequence(int i) : Fsequence()
{
  if (i>0){
    _deb=new Cdescripteur[i];
    _vfin=_fin=&_deb[i-1];
  }    
}

Csequence::Csequence(void*) : Fsequence()
{}

Csequence::Csequence(const Fsequence& g) 
{
  cerr << "Bad call of Csequence::Csequence(const Fsequence& g) ";
  cerr << endl;
  exit(0);
}

void Csequence::genere(int l){
  nettoie();
  if (l>0){
    _deb=new Cdescripteur[l];
    _vfin=_fin=&_deb[l-1];
  }
}
  
void Csequence::nettoie()
{
  if (_deb)
    delete[] _deb;
  _deb=0;
}

Csequence::~Csequence()
{
  nettoie();
}

/////////////////////////////////////////////////////
/// modifs

void Csequence::egale(const Fsequence& G)
{
  this->operator=(G);
}

void Csequence::operator=(const Fsequence& G)
{
  int lG=G.taille();
  if (taille()!=lG){
    if (_deb)
      delete[] _deb;
    if (!lG)
      return;
    _deb=new Cdescripteur[lG];
    _vfin=_fin=_deb+lG-1;
  }
  int i;
  for (i=0;i<lG;i++)
    *(_deb+i)=*(G.deb()+i);
  }

////////////////////////////////////////////
// recuperations de fichiers
////////////////////////////////////////////

bool Csequence::recup_seq(const string& st)
{
  int lg;
  Cdescripteur *p_c;
  char c;
  int deb;
  int i,j;
  char buffer[12];
  
  ifstream flot(st.c_str());
  if (!flot){
    cout << "pb opening " << st << endl;
    return 0;
  }
	
  flot >> c;
  i=0;
  while (isdigit(c) && (i<11)){
    buffer[i++]=c;
    flot.get(c);
  }
  buffer[i]='\0';
  deb=atoi(buffer);

  if (deb<=0){
    cout << "bad start in file ";
    cout << st << endl;
    flot.close();
    return 0;
  }

  flot >> c;
  while (!isgraph(c))
    flot.get(c);
  if (isdigit(c)){
    i=0;
    while (isdigit(c) && (i<11)){
      buffer[i++]=c;
      flot.get(c);
    }
    buffer[i]='\0';
    if ((lg=1+atoi(buffer)-deb)<=0){
      cout << "bad boundaries in file ";
      cout << st << endl;
      flot.close();
      return 0;
    }
    while (!isgraph(c))
      flot.get(c);
  }
  else{
    lg=deb;
    deb=1;
  }

  ////////// recuperation des elements de la sequence

  if (_deb)
    delete[] _deb;
  _deb=new Cdescripteur[lg];
  _vfin=_fin=&(_deb[lg-1]);

  p_c=_deb;

  //  on va jusqu'a deb sans rien faire

  i=1;j=0;
  while (!flot.eof() && (i<deb) && (j<100)){
    if (isalpha(c)){
      i++;
      j=0;
    }
    else 
      j++;
    flot.get(c);
  }
  if (flot.eof()){
    cout << "Bad boundaries in file ";
    cout << st << endl;
    if (_deb)
      delete[] _deb;
    _vfin=_fin=_deb=0;
    flot.close();
    return 0;
  }
  if (j==100){
    cout << "Pb in file ";
    cout << st << endl;
    if (_deb)
      delete[] _deb;
    _vfin=_fin=_deb=0;
    flot.close();
    return 0;
  }

  // puis on remplit la sequence

  while (!flot.eof() && (c!='.') && (p_c<=_fin)){
    if (isalpha(c)){
      *p_c=c;
      p_c++;
    }
    flot.get(c);
  }

  if (p_c<=_fin){
    cout << "Sequence-length pb" << endl;
    _fin=--p_c;
  }
  
  while (!flot.eof() && (c!='.'))
    flot >> c;

  flot.close();
  return 1;
}

bool Csequence::recup_fst(const string& st)
{
  ifstream flot(st.c_str());
  if (!flot){
    cout << "pb opening " << st << endl;
    return 0;
  }
  recup_fst(flot,0);
  flot.close();
  return 1;
}    


int Csequence::recup_fst(ifstream& flot, int deb)
{
  long lg;
  int pos, fin;
  char c;
  Cdescripteur* p_c;
  flot.seekg(0,iostream::end);
  fin=flot.tellg();
  fin--;
  if (deb>=fin)
    return -1;
  flot.seekg(deb,iostream::beg);
  
  flot.ignore(1000,'\n');
  lg=0;
  
  c=flot.peek();
  while ((c!='>') && (fin>flot.tellg())){
    if (isalpha(c) || c=='-')
      lg++;
    flot.seekg(1,iostream::cur);
    c=flot.peek();
  }
  if (isalpha(c) || c=='-')
    lg++;
  
  pos=flot.tellg();
  // on recommence
  
  if (_deb)
    delete[] _deb;
  
  if (lg==0){
    cout << "Empty sequence" << endl;
    return -2;
  }
  if (!(_deb=new Cdescripteur[lg])){
    cout << "Allocation pb in Csequence::recup_fst()." << endl;
    return -2;
  }
  
  _vfin=_fin=&(_deb[lg-1]);
  
  flot.seekg(deb,iostream::beg);
  flot.ignore(1000,'\n');
  
  flot >> c;
  p_c=_deb;
  while ((fin>=flot.tellg()) && (p_c<_fin)){
    if (isalpha(c) || c=='-'){
      *p_c=c;
      p_c++;
    }
    flot >> c;
  }
  if (isalpha(c) || c=='-'){
    *p_c=c;
    p_c++;
  }
  
  if (p_c<=_fin){
    cerr << "Memory waste in sequence" << endl;
    _fin=p_c;
  }

  return (pos==fin)?-1:pos;
}
		
bool Csequence::recup_rel(const Fsequence& F, const Csegment& S)
{
  nettoie();

  int d,f;
  int Ft=F.taille();
  
  d=S.pos_d();
  f=S.pos_f();
  if (f>Ft)
    f=Ft;
  
  if ((!(F.deb())) || (d > f))
    return 0;

  _deb=new Cdescripteur[f-d+1];
  _fin=_deb;
  Cdescripteur *po, *pf;
  po=&F[d-1];
  pf=&F[f-1];

  do{
    *(_fin++)=*(po++);
  } while (po<=pf);
  _vfin=--_fin;
  
  return 1;
}

File: sequence.py
# Copyright 2004 Laurent GUEGUEN
# 
# This file is part of Sarment.
# 
# Sarment is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# Sarment is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with Sarment; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from __future__ import absolute_import
from __future__ import print_function
import random 
import string
import math

from Modules.Csequence import *
from Modules.Fsequence import *

import partition
from segment import Segment
from six.moves import range
from functools import reduce

class _Seq:
    "virtual Sequence"

    def __init__(self):
        self.__gen=Fsequence()
        self.__nom=""
        
    def __c_elem(self):
        "Return C++ object"
        return self.__gen
    

#########################################
### modifications

    def g_name(self, s):
        "Take name s"
        self.__nom=s

    def __setitem__(self, i, c):
        "(x.__setitem__(i,c) <==> x[i]=c)"
        self.__gen[i]=c

    def __setslice__(self, i, j, s):
        """(x.__setslice__(i,j,s) <==> x[i:j]=s)
Replace the subsequence x[i:j] by Sequence or string s.
If len(s) <> j-i, builds a new Sequence as a result,
otherwise replace in the former one.
"""
        if i<0:
            i=0
        if j>len(self):
            j=len(self)
        ls=len(s)
        if ls==j-i:
            for k in range(i,j):
                self[k]=s[k-i]
        else:
            g2=Csequence()
            lg2=len(self)-(j-i)+ls
            g2.genere(lg2)            
            k=0
            while k<i:
                g2[k]=self[k]
                k+=1
            while k<i+ls:
                g2[k]=s[k-i]
                k+=1
            while k<lg2:
                g2[k]=self[k-ls+(j-i)]
                k+=1
            self.__gen=g2
            
    def shuffle(self,i=0):
        """Quasi-uniformly randomize itself using i (default:
len*(log(len)+1)/2) transpositions.
"""
        if i==0:
            i=(int)(len(self)*(math.log(len(self))+1)/2)

        self.__gen.melange(i)

    def read_Lprop(self, lprop, **kw):
        """Build from a Lproportion lprop. Return the Partition of the
descriptors.
        
Keyword argument:
deb -- change only after position deb (>=0) included;
fin -- change only before position fin (<len()) included;
long -- create a lg-length Sequence. In that case, deb and
        fin are not read;
etat_init --  begin with descriptor number etat_init if it is valid.
              Otherwise, starts with a random descriptor of lprop.
"""

        al=lprop.alph()
        if al==[] or al==["^"]:
            print("Empty Lproportion")
            self.__init__()
            return

        p=partition.Partition()

        e=kw.get('etat_init',-1)
    
        if 'long' in kw:
            if not isinstance(self, Sequence):
                print("impossible generation of virtual sequence")
                return p
            fin=kw['long']
            if fin<=0:
                return p
            self.generate(fin)
            fin-=1
            deb=0
        else:
            self.__gen.termine(self.__gen.vtaille()-1)
            fin=kw.get('fin',len(self)-1)
            fin=min(fin,len(self))
            deb=kw.get('deb',0)
            if deb<0 or deb>fin:
                return p

        lik=lprop.num()
        if lik==[]:
            return p

        if deb!=0:
            p+=Segment(deb=0,fin=deb-1)
            
        def sum(x,y):
            return x+y

        # dicos des transitions entre etats
        ltrans={}
        llik=len(lik)
        for b in lik:
            sb=reduce(sum, [lprop.inter(b,x) for x in lik])
            ltrans[b]={}
            if sb==0:
                for x in lik:
                    ltrans[b][x]=1.0/llik
            else:
                for x in lik:
                    ltrans[b][x]=float(lprop.inter(b,x))/sb
                    
        ## Demarrage avec choix uniforme de l'etat initial ou bien e        

        if e in lik:
            etat=e
        else:
            etat=random.choice(lik)

        tab=lprop[etat]
        
        dseg=deb
        ilal=1.0/len(al)
        asuiv=[[x,ilal] for x in al]
        

        
        for i in range(deb,fin):

            ##choix de la lettre

            j=i-1
            s=""
            while j>=0:
                s=self.__gen[j]+s
                if not tab.has_prefix(s):
                    s=s[1:]
                    break
                else:
                    j-=1

            suiv=tab.next(s)
            if suiv==[]:
                suiv=asuiv

            y=reduce(sum ,[t[1] for t in suiv])
            x=random.random()*y

            lsuiv=len(suiv)            
            while x>=0 and lsuiv>0:
                lsuiv-=1
                x-=suiv[lsuiv][1]

            if len(suiv[lsuiv][0])!=1:
                print("Bad proportion after " + sq[1:])
                self.__init__()
                return
            
            a=suiv[lsuiv][0][0]
            if a=="^":
                for j in range(i,fin+1):
                    self.__gen[j]="\0"
                self.__gen.termine(i-1)
                if dseg<i:
                    p+=Segment(deb=dseg,fin=i-1,num=[etat])
                return p
            else: 
                self.__gen[i]=a

            ##choix de l'etat suivant

            llik=len(lik)
            x=random.random()
            while x>=0 and llik>0:
                llik-=1
                x-=ltrans[etat][lik[llik]]

            if etat!=lik[llik]:
                p+=Segment(deb=dseg,fin=i,num=[etat])
                dseg=i+1
                etat=lik[llik]
                tab=lprop[etat]


        ##choix de la lettre finale

        j=fin-1
        s=''
        while j>=0:
            s=self.__gen[j]+s
            if not tab.has_prefix(s):
                s=s[1:]
                break
            else:
                j-=1
                
        suiv=tab.next(s)
        if suiv==[]:
            suiv=asuiv

        y=reduce(lambda x,y: x+y ,[t[1] for t in suiv])
        x=random.random()*y
        
        lsuiv=len(suiv)            
        while x>=0 and lsuiv>0:
            lsuiv-=1
            x-=suiv[lsuiv][1]

        a=suiv[lsuiv][0][0]
        if a=="^":
            self.__gen.termine(fin-1)
            self.__gen[fin]="\0"
            if dseg<fin:
                p+=Segment(deb=dseg,fin=fin-1,num=[etat])
            return p
        else: 
            self.__gen[fin]=a
            p+=Segment(deb=dseg,fin=fin,num=[etat])

        if fin<len(self)-1:
            p+=Segment(deb=fin+1,fin=len(self)-1)
            
        return p


    def read_prop(self, prop, **kw):
        """Build from a Proportion prop. 

Keyword argument:
deb -- change only after position deb (>=0) included;
fin -- change only before position fin (<len()) included;
long -- create a lg-length Sequence. In that case, deb and
        fin are not read.
"""


        al=prop.alph()
        if al==[] or al==['^']:
            print("Empty Proportion")
            self.__init__()
            return

        asuiv=[[x,1.0/len(al)] for x in al]
        
        if 'long' in kw:
            if not isinstance(self, Sequence):
                print("impossible generation of virtual sequence")
                return p
            fin=kw['long']
            if fin<=0:
                return
            self.generate(fin)
            fin-=1
            deb=0
        else:
            self.__gen.termine(self.__gen.vtaille()-1)
            fin=kw.get('fin',len(self)-1)
            fin=min(fin,len(self)-1)
            deb=kw.get('deb',0)
            if deb<0 or deb>fin:
                return
            
        def sum(x,y):
            return x+y

        for i in range(deb,fin+1):
            j=i-1
            s=""
            while j>=0:
                s=self.__gen[j]+s
                if not prop.has_prefix(s):
                    s=s[1:]
                    break
                else:
                    j-=1

            suiv=prop.next(s)
            if j==-1:
                suiv+=prop.next("^"+s)

            if suiv==[]:
                suiv=asuiv

            y=reduce(sum ,[t[1] for t in suiv])
            x=random.random()*y
            lsuiv=len(suiv)
            
            while x>=0 and lsuiv>0:
                lsuiv-=1
                x-=suiv[lsuiv][1]

            a=suiv[lsuiv][0][0]
            if a=="^":
                self.__gen.termine(i-1)
                for j in range(i,fin+1):
                    self.__gen[i]="\0"
                break
            else: 
                self.__gen[i]=a

###########################################
### donnees

    def name(self):
        "Return the name."
        return self.__nom
            
    def __getitem__(self, i):
        "(x.__getitem__(i) <==> x[i])"
        return self.__gen[i]

    def __len__(self):
        "(x.__len__() <==> len(x))"
        return len(self.__gen)

    def alpha(self):
        "Return the list of the used letters."
        st=[]
        for i in self:
            if not i in st:
                st.append(i)
        return st

    def __getslice__(self,i,j):
        """!!! Do NOT create a new Sequence; so destruction of the original
Sequence entails destruction of this result.
(x.__getslice__(i,j) <==> x[i:j])"""
        nv=_Seq()
        nv.__c_elem().recup_rel(self.__c_elem(),i,j-1)
        nv.g_name(self.name())
        return nv

    def copy(self):
        "Return a NEW copy Sequence of the Sequence."
        g=Sequence()
        g.__gen.copie(self.__c_elem())
        g._Seq__nom=self._Seq__nom
        return g

############################################
###   sorties

    def __str__(self):
        """(x.__str__()<==> str(x))"""
        return str(self.__gen)
    
    def fasta(self, lg=80):
        "Return the string in fasta format."
        s=">" + str(self.__nom) + '\n'
        j=0
        if lg==0:
            lg=len(self)
        if len(self)>0:
            for i in self:
                s+=i
                j+=1
                if j==lg:
                    s+="\n"
                    j=0
            s+="\n"
        return s
	
    def seq(self):
        "Return the mere sequence."
        s=""
        j=0
        if len(self)>0:
            for i in self:
                s+=i
        return s

    


#######################################################################
#######################################################################
########  Classe Sequence

class Sequence(_Seq):
    "Sequence"

    def __init__(self, **kw):
        """Keyword argument:
fic -- build from filename fic.
"""
        _Seq.__init__(self)
        self._Seq__gen=Csequence()
        if 'fic' in kw:
            self.read_nf(kw['fic'])

    def generate(self,l):
        "Build empty with length l."
        self._Seq__gen.genere(l)

    def read_nf(self,nf):
        "Build from filename nf."
        self._Seq__gen.read_nf(nf)
        if nf[-3:]==".fa" or nf[-4:]==".fst" or nf[-6:]==".fasta":
            try:
                f=open(nf,'rU')
            except IOError:
                return
            
            l=f.readline()
            while l:
                if l[0]=='>':
                    self._Seq__nom=l[1:]
                    break
                l=f.readline()
            f.close()

#########################################
### modifications
    
    def read_Part(self, part, lprop):
        """Compute randomly from a Partition part and a Lproportion
lprop. Segments of part must be numbered, which numbers are those of
descriptors of lprop.
"""

        al=lprop.alph()
        if al==[] or al==['^']:
            print("Empty Lproportion")
            self.__init__()
            return

        asuiv=[[x,1.0/len(al)] for x in al]
        
        lg=part.len_don()
        if lg<=0:
            return
        ln=part.num()
        ln2=lprop.num()
        for i in ln:
            if not i in ln2:
                print("Bad Lproportion: ",i," not in the Lproportion")
                return

        self.generate(lg)

        for s in part:
            letat=[lprop[i] for i in s.num()]
            lle=len(letat)
            i=0
            for pos in range(s.deb(),s.fin()+1):
                tab=letat[i]

                ##choix de la lettre

                j=pos-1
                sq=""
                suiv=[]
                while j>=0:
                    sq=self._Seq__gen[j]+sq
                    if not tab.has_prefix(sq):
                        sq=sq[1:]
                        break
                    else:
                        j-=1

                suiv=tab.next(sq)                        
                if j==-1:
                    suiv+=tab.next("^"+sq)

                if suiv==[]:
                    suiv=asuiv

                x=random.random()
                lsuiv=len(suiv)
                while x>=0 and lsuiv>0:
                    lsuiv-=1
                    x-=suiv[lsuiv][1]

                if len(suiv[lsuiv][0])!=1:
                    print("Bad proportion after " + sq[1:])
                    self.__init__()
                    return

                if suiv[lsuiv][0]=='^':
                    for j in range(pos,lg):
                        self._Seq__gen[j]="\0"
                    self._Seq__gen.termine(pos-1)
                    return
                
                self._Seq__gen[pos]=suiv[lsuiv][0]
                
                i+=1
                if i>=lle:
                    i=0
